src/BenchmarkDotNet/Diagnosers/EventPipeProfiler.cs

```diff
@@ -92,144 +76,78 @@ public void Handle(HostSignal signal, DiagnoserActionParameters parameters)
-                            if (nBytesRead <= 0)
-                                break;
-                            fs.Write(buffer, 0, nBytesRead);
-                        }
+                        fs.Write(buffer, 0, bytesRead);
                     }
+
+                    fs.Flush();
                 }
-                catch (Exception ex)
-                {
-                    logger.WriteLine(LogKind.Error, $"An exception occurred during reading trace stream: {ex}");
-                }
-            });
-            collectingTask.Start();
+            }
+            catch (Exception ex)
+            {
+                logger.WriteLine(LogKind.Error, $"An exception occurred during reading trace stream: {ex}");
+            }
         }
 
         public IEnumerable<Metric> ProcessResults(DiagnoserResults results)
         {
             Task.WaitAll(collectingTask);
 
             if (benchmarkToTraceFile.TryGetValue(results.BenchmarkCase, out var traceFilePath))
-                ConvertToSpeedscope(results.BenchmarkCase, traceFilePath);
+                benchmarkToTraceFile[results.BenchmarkCase] = SpeedScopeExporter.Convert(traceFilePath, results.BenchmarkCase.Config.GetCompositeLogger());
 
             return Array.Empty<Metric>();
         }
-        
+
         public void DisplayResults(ILogger resultLogger)
         {
-            resultLogger.WriteLine();
-            resultLogger.WriteLineHeader(LogSeparator);
-
-            if (logger.CapturedOutput.Any())
-            {
-                foreach (var line in logger.CapturedOutput)
-                    resultLogger.Write(line.Kind, line.Text);
-
-                resultLogger.WriteLine();
-            }
             if (!benchmarkToTraceFile.Any())
                 return;
 
             resultLogger.WriteLineInfo($"Exported {benchmarkToTraceFile.Count} trace file(s). Example:");
             resultLogger.WriteLineInfo(benchmarkToTraceFile.Values.First());
-
-            resultLogger.WriteLineHeader(LogSeparator);
-        }
-
-        public IEnumerable<ValidationError> Validate(ValidationParameters validationParameters)
-        {
-            foreach (var benchmark in validationParameters.Benchmarks)
-            {
-                var runtime = benchmark.Job.ResolveValue(EnvironmentMode.RuntimeCharacteristic, EnvironmentResolver.Instance);
-
-                if (runtime.RuntimeMoniker < RuntimeMoniker.NetCoreApp30)
-                {
-                    yield return new ValidationError(true, $"{nameof(EventPipeProfiler)} supports only .NET Core 3.0+", benchmark);
-                }
-            }
         }
 
-        public string ShortName => "EventPipe";
-
-        private void ConvertToSpeedscope(BenchmarkCase benchmarkCase, string traceFilePath)
+        private static ImmutableHashSet<EventPipeProvider> MapToProviders(EventPipeProfile profile, IReadOnlyCollection<EventPipeProvider> providers)
```

Comment:
This function has a little different behavior than my old code and the `dotnet trace` tool. 

There you can find the documentation of `dotnet trace`: https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-trace

>* --providers \<list-of-comma-separated-providers>
> A comma-separated list of EventPipe providers to be enabled. These providers supplement any providers implied by --profile <profile-name>. If there's any inconsistency for a particular provider, this configuration takes precedence over the implicit configuration from the profile.

So for example a user could send:
1. profile = `EventPipeProfile.CpuSampling`
2. providers = `new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Verbose, (long) ClrTraceEventParser.Keywords.Default)`

```c#
 new EventPipeProfiler(EventPipeProfile.CpuSampling, new [] { new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Verbose, (long) ClrTraceEventParser.Keywords.Default | ClrTraceEventParser.Keywords.JitTracing)});
```

This should give all the providers sent by the user and all providers resulting from the profile except those that were sent in the providers.

In the above example it should use:

1. sent by user - `new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Verbose, (long) ClrTraceEventParser.Keywords.Default | ClrTraceEventParser.Keywords.JitTracing)`
2. from the profile - `new EventPipeProvider("Microsoft-DotNETCore-SampleProfiler", EventLevel.Informational)`

This from the profile won't be used:
`new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Informational, (long) ClrTraceEventParser.Keywords.Default)`

Reply:
@WojciechNagorski great catch! I've added tests for that and fixed it. PTAL again

https://github.com/dotnet/BenchmarkDotNet/pull/1381/commits/68a207d6c3627be3372de6370d17461f6d2a0c6b